File: react-native/library/src/starPlanet/components/TspChildSwitcher.js
import React from 'react';
import { Image, Pressable, ScrollView, StyleSheet, Text, View } from 'react-native';
import { withTheme } from '../utils/shared';

/**
 * TspChildSwitcher – Single-select child chips/tabs.
 * Item label resolves from label | name | text.
 */
export function TspChildSwitcher({
  items = [],
  selectedId,
  variant = 'chip',
  disabled = false,
  theme,
  onChange,
}) {
  const t = withTheme(theme);
  const useTabs = variant === 'tabs';

  const row = items.map((item) => {
    const id = item?.id;
    const label = item?.label ?? item?.name ?? item?.text ?? '';
    const selected = String(id) === String(selectedId);
    const leading = item?.leading
      ? item.leading
      : item?.iconSrc
        ? (
          <Image
            source={{ uri: item.iconSrc }}
            style={styles.avatar}
            accessibilityIgnoresInvertColors
          />
        )
        : item?.emoji
          ? <Text style={styles.emoji}>{item.emoji}</Text>
          : null;

    return (
      <Pressable
        key={String(id)}
        disabled={disabled}
        onPress={disabled ? undefined : () => onChange?.(id)}
        style={[
          useTabs ? styles.tab : styles.chip,
          useTabs && styles.tabFlex,
          {
            backgroundColor: selected ? t.brandPrimary : t.surfaceRaised,
            borderColor: selected ? t.brandPrimary : t.borderDefault,
          },
          disabled && styles.disabled,
        ]}
        accessibilityRole="tab"
        accessibilityState={{ selected, disabled }}
      >
        {leading}
        <Text
          style={[styles.label, { color: selected ? '#FFFFFF' : t.textSecondary }]}
          numberOfLines={1}
        >
          {label}
        </Text>
      </Pressable>
    );
  });

  if (useTabs) {
    return <View style={styles.row}>{row}</View>;
  }

  return (
    <ScrollView horizontal showsHorizontalScrollIndicator={false} contentContainerStyle={styles.row}>
      {row}
    </ScrollView>
  );
}

const styles = StyleSheet.create({
  row: {
    flexDirection: 'row',
    alignItems: 'center',
    gap: 8,
    paddingVertical: 4,
  },
  chip: {
    minHeight: 44,
    paddingHorizontal: 14,
    borderRadius: 999,
    borderWidth: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    gap: 8,
  },
  tab: {
    minHeight: 44,
    paddingHorizontal: 14,
    borderRadius: 14,
    borderWidth: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    gap: 8,
  },
  tabFlex: { flex: 1 },
  avatar: {
    width: 22,
    height: 22,
    borderRadius: 8,
    backgroundColor: '#FFFFFF',
  },
  emoji: { fontSize: 16 },
  label: { fontWeight: '800', fontSize: 14 },
  disabled: { opacity: 0.45 },
});

File: flutter/library/lib/src/tsp_child_switcher.dart
import 'package:flutter/material.dart';
import 'star_planet_theme.dart';

class TspChildSwitcherItem {
  const TspChildSwitcherItem({
    required this.id,
    this.label,
    this.name,
    this.text,
    this.iconSrc,
    this.emoji,
    this.leading,
  });

  final Object id;
  final String? label;
  final String? name;
  final String? text;
  final String? iconSrc;
  final String? emoji;
  final Widget? leading;

  String get resolvedLabel => label ?? name ?? text ?? '';
}

/// Single-select child chips/tabs.
class TspChildSwitcher extends StatelessWidget {
  const TspChildSwitcher({
    super.key,
    this.items = const [],
    this.selectedId,
    this.variant = 'chip',
    this.disabled = false,
    this.theme = StarPlanetTheme.sky,
    this.onChanged,
  });

  final List<TspChildSwitcherItem> items;
  final Object? selectedId;
  final String variant;
  final bool disabled;
  final StarPlanetTheme theme;
  final ValueChanged<Object>? onChanged;

  @override
  Widget build(BuildContext context) {
    final useTabs = variant == 'tabs';
    final row = Row(
      children: [
        for (var i = 0; i < items.length; i++) ...[
          if (i > 0) const SizedBox(width: 8),
          Flexible(
            fit: useTabs ? FlexFit.tight : FlexFit.loose,
            child: _buildItem(items[i], useTabs),
          ),
        ],
      ],
    );

    if (useTabs) return row;

    return SingleChildScrollView(
      scrollDirection: Axis.horizontal,
      padding: const EdgeInsets.symmetric(vertical: 4),
      child: Row(
        children: [
          for (var i = 0; i < items.length; i++) ...[
            if (i > 0) const SizedBox(width: 8),
            _buildItem(items[i], false),
          ],
        ],
      ),
    );
  }

  Widget _buildItem(TspChildSwitcherItem item, bool useTabs) {
    final selected = '${item.id}' == '$selectedId';
    final leading = item.leading ??
        (item.iconSrc != null
            ? ClipRRect(
                borderRadius: BorderRadius.circular(8),
                child: Image.network(item.iconSrc!, width: 22, height: 22, fit: BoxFit.cover),
              )
            : item.emoji != null
                ? Text(item.emoji!, style: const TextStyle(fontSize: 16))
                : null);

    return Opacity(
      opacity: disabled ? .45 : 1,
      child: GestureDetector(
        onTap: disabled ? null : () => onChanged?.call(item.id),
        child: Container(
          constraints: const BoxConstraints(minHeight: 44),
          padding: const EdgeInsets.symmetric(horizontal: 14),
          decoration: BoxDecoration(
            color: selected ? theme.brandPrimary : theme.surfaceRaised,
            borderRadius: BorderRadius.circular(useTabs ? 14 : 999),
            border: Border.all(color: selected ? theme.brandPrimary : theme.borderDefault),
          ),
          child: Row(
            mainAxisSize: useTabs ? MainAxisSize.max : MainAxisSize.min,
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              if (leading != null) ...[leading, const SizedBox(width: 8)],
              Flexible(
                child: Text(
                  item.resolvedLabel,
                  maxLines: 1,
                  overflow: TextOverflow.ellipsis,
                  style: TextStyle(
                    color: selected ? Colors.white : theme.textSecondary,
                    fontWeight: FontWeight.w800,
                    fontSize: 14,
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
